instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt

```diff
@@ -281,19 +273,195 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
+            profile = DirectionsCriteria.PROFILE_DRIVING_TRAFFIC,
+            jsonResponse = mockRoute.routeResponseJson,
+            expectedCoordinates = mockRoute.routeWaypoints,
+            relaxedExpectedCoordinates = false
+        )
+        mockWebServerRule.requestHandlers.add(requestHandler)
+        val rerouteRequestHandler = MockDirectionsRequestHandler(
+            profile = DirectionsCriteria.PROFILE_DRIVING_TRAFFIC,
+            jsonResponse = mockRouteReroute.routeResponseJson,
+            expectedCoordinates = listOf(
+                offRouteLocation,
+                mockRoute.routeWaypoints.last()
             ),
-            rerouteStates
+            relaxedExpectedCoordinates = true
         )
+        val responseModifier = DelayedResponseModifier(4000) { it }
+        rerouteRequestHandler.jsonResponseModifier = responseModifier
+        mockWebServerRule.requestHandlers.add(rerouteRequestHandler)
+        val refreshHandler = MockDirectionsRefreshHandler(
+            testUuid = "jpCHHUC26qFOwISCNLjar2xmTfI6Dxd0qCHOoqwt_1VAlESNvsr7Zg==",
+            readRawFileText(
+                activity,
+                R.raw.route_response_dc_short_with_alternative_refresh_route_0
+            ),
+            acceptedGeometryIndex = 0
+        )
+        mockWebServerRule.requestHandlers.add(refreshHandler)
+
+        val rerouteStateTransitionAssertion = RerouteStateTransitionAssertion(
+            mapboxNavigation.getRerouteController()!!
+        ) {
+            requiredState(RerouteState.Idle)
+            requiredState(RerouteState.FetchingRoute)
+            requiredState(RerouteState.RouteFetched(RouterOrigin.Offboard))
+            requiredState(RerouteState.Idle)
+        }
+
+        val originalRoutes = mapboxNavigation.requestRoutes(
+            RouteOptions.builder()
+                .applyDefaultNavigationOptions()
+                .applyLanguageAndVoiceUnitOptions(activity)
+                .baseUrl(mockWebServerRule.baseUrl)
+                .coordinatesList(mockRoute.routeWaypoints)
+                .alternatives(true)
+                .build()
+        ).getSuccessfulResultOrThrowException().routes
+        mapboxHistoryTestRule.historyRecorder.startRecording()
+        mockLocationReplayerRule.loopUpdate(originLocationUpdate, times = 120)
+        mapboxNavigation.startTripSession()
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(listOf(originalRoutes.first()))
+
+        mapboxNavigation.routeProgressUpdates().filter {
+            it.currentState == RouteProgressState.TRACKING
+        }.first()
+        mockLocationReplayerRule.stopAndClearEvents()
+        mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 120)
+        // wait for OFF_ROUTE
+        mapboxNavigation.offRouteUpdates().filter { it }.first()
+        // wait for refresh update
+        mapboxNavigation.routesUpdates().first {
+            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH
```

Comment:
Aren't you afraid that the reroute will finish before you register the observer? Won't it break anything? I mean you have:
```
mapboxNavigation.routesUpdates().first {
            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH
}
val routesUpdate = mapboxNavigation.routesUpdates().first {
            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
}
```
I'm worried one of the line might get stuck.

Reply:
The refresh delay in the test is 3 seconds, the reroute 4 seconds, so I'm expecting the refresh to happen before the reroute. If that wasn't the case, this code would indeed get stuck and time out but that's good as it would give us the visibility into the fact the test is not testing what it's meant to be testing anymore.

Unfortunately, I have only 5 seconds to work with because that's the default timeout in Nav Native. We can refactor that in the future when we expose the granular request timeouts option

Reply:
Ah, now I see. Didn't pay attention to the intervals and delays. Just for clarity it would be nice to add something like:
```
// refreshInterval has to be less than the reroute delay as we expect the refresh to happen before the reroute
val refreshInterval = 3000
// reroute delay has to be less than 5 seconds as it's the default timeout in NN
val rerouteDelay = 4000
```